### PR TITLE
Add support for GICv3 virtualization on aarch64

### PR DESCRIPTION
**Pull request by bennoleslie**
This adds sufficient kernel support for the GICv3 interrupt controller
to be used in a virtualization context on aarch64.

This set of changes has some limitations, however it is still an
improvement on the status quo.

Limitations:

1: This only provides support for aarch64. Anyone wanting support
for aarch32 + GICv3 + virtualization would need to add additional
code.

2: This code only supports 32 priority levels. Support for more
than 32 priority requires changing the get/set_gic_vcpu_ctrl_apr
interface. This is feasible, but requires a more invasive set of
changes. 32 priority levels has been shown to be sufficient in
practise.

Impacts on verification:

This set of changes should only impact Aarch64 Hypervisor
configurations. This is not yet verified so should not have
an impact on verification.

Level of testing:

This has been tested on an iMX8QXP based board. Testing
has at this point in time been limited to a single virtual
machine.

Note: support for this board is not yet upstrea, but is
currently being prepared.

Explanation of changes:

Ideally a new config item would not be required and this
could be driven purely by DTS and hardware.yml configuration.
However, the structures.bf requires changes. This can only
deal with config.h header files, not other more complex
header files. As such it was necessary to introduce a config
item which can be used for this purpose.

The appropriate platforms (as determined by examination of
DTS files) have been updated with the appropriate config
setting. This config setting only has any relevance if
hypervisor mode is already enabled, so should not cause
any difficulty for existing code or configuration.

Note: No testing has been performed on the updated
platforms.

There may be alternative factorings of this, which could
be considered in future work.

Signed-off-by: Ben Leslie <benno@brkawy.com>

### CODE REVIEW

**Comment by kent-mcleod** [comment]
@bennoleslie, are you planning to make the follow-up PRs as the changes become available or only one at a time?

Also, my recollection was that (https://github.com/seL4/seL4/pull/145) was blocked until there was working user level examples that could be placed into CI somewhere.  But as far as I know work on this has been stalled for quite a while.  Do you have any suggestions on how we could set up CI for these new changes?

**Review by Indanz — COMMENTED** [review]
The code seems a straightforward adaption of the current GICv2 code to support GICv3 virtualization too. I do not see any  issues that could cause existing code to break.

I think it would be good to mention what relation this implementation has to the #145 one though. At first glance it seems there is no code-reuse, but do any of the concerns of @yyshen and questions of @pingerino apply to this version?

Is there a working user level example in CI that tests the current GICv2 virtualization support?
If not, it seems unreasonable to demand the same from this pull request.

We are using this patchset to run a Linux VM on an i.MX8 board in development, if there are any issues they will be found and fixed.

**Comment by Indanz** [comment]
What needs to be done to get this merged upstream?

**Comment by bennoleslie** [comment]
I've remerged this to the latest master.

I'm not sure what the best way to address this being tested in CI would be.

Per the comments this has been tested on an iMX8QXP board with a virtualised Linux, but I'm not sure how that could be added to CI.

I need to get the rest of the patches that support that board merged into master, but adding an seL4Test test case would be a significant amount of work that I'm afraid I can't commit to at this time.

I've been trying to add the patches required to support this board (and use case) one at a time so that they are in easy to digest chunks for both myself and reviewers. I find that adding all at once causes confusion in practise.

Assuming all existing tests pass, this doesn't introduce any regression so I think leaves the code base in a better place than not having it. What exactly would it take to get it merged?

Note: Indan's comments are valid, but I was expressly trying to avoid modifying / refactoring any existing code to avoid any chance of introducing a regression. I'd prefer to make that refactor as a follow-on PR after this is merged if possible

**Review by lsf37 — APPROVED** [review]
I'd be happy to merge this now and keep working on refactoring things incrementally. There is a chance we'll be getting a iMX8QXP board into the test as well.

As long as we don't mark this as "supported" yet, it doesn't break anything existing, and there is some commitment to keep this working and get it completed, I'm fine with the incremental approach.

@kent-mcleod any concerns from your side before I go ahead?

**Comment by lsf37** [comment]
(can confirm there is no verification impact)

**Comment by bennoleslie** [comment]
Note: #520 adds board support for a target on which this code can run.

**Comment by kent-mcleod** [comment]
If we aren't planning on requiring user level tests before merging this feature, then couldn't we also consider merging #145 instead as it goes further in integrating the changes with the rest of the GIC virtualization code? For example, there are several changes in src/arch/arm/object/vcpu.c which remove GICv2 specific assumptions and generalize it to also be able to use GICv3 (https://github.com/seL4/seL4/pull/145/commits/54fd3de7de8a03755f37586607b9263377d96031).   At the end of last year data61 updated the old PR to current master and there's no conflicts updating it to current master (https://github.com/seL4/seL4/commits/sylvain/gic-v3) and sel4test passes on imx8mm with virtualization enabled.  Of course if it just doesn't work at the moment, then I'd be happy for us to close the old PR and merge the current PR if that's what everyone prefers. 

We would be able to rely on the existing GICv2 CI to catch any obvious issues in changes to the GICv2 that are unintentionally introduced (the PR does try to keep the existing GICv2 implementation the same but there do appear to be some changes (https://github.com/seL4/seL4/pull/145/files#diff-23bfdc3e8929825739e6748453759955ad63724cce935ad72dd91b17e364db8eR423).

> Is there a working user level example in CI that tests the current GICv2 virtualization support?
> If not, it seems unreasonable to demand the same from this pull request.

At the time when  #145  was made, there was working CI that tested the virtualization capabilities of the GICV2 via the camkes-vm-examples project ( https://github.com/seL4/camkes-vm-examples/tree/master/apps/Arm ). Generally the following features were being integration tested:
- That the kernel implementation could be successfully combined with a user level hypervisor implementation that could be used to boot a Linux vm indicating that both the hardware virtualization and the software emulation parts of the GICv2 would somewhat work.
- That interrupts could be injected for passthrough devices and emulated devices and the guest could handle them
- That when running multiple VMs, context switching would work where the VGIC state needed to be saved and restored correctly by the kernel, and that IRQs arriving when a VM was inactive would still be delivered when it resumed.
- That the GICv2 virtualization worked on more than a single platform or Arm CPU version. (E.G. the exynos5410 also needed an IRQ combiner to be emulated to support passthrough devices).

I'm not sure what the current state of the CI is though after the recent rapid migration. @lsf37 would know though?

**Comment by bennoleslie** [comment]
I assume the pc99 failure is spurious in this case.

**Comment by bennoleslie** [comment]
> If we aren't planning on requiring user level tests before merging this feature, then couldn't we also consider merging #145 instead as it goes further in integrating the changes with the rest of the GIC virtualization code? 

I don't have a strong feeling one way or the other. I've got zero problem ditching this PR if someone wants to do the work to get #145 ready for merge. I've also got zero problems with getting this merged now, and then applying the refactors that are in #145.

I think getting *something* in sooner rather than later is beneficial as it allows others to build on this (including potentially getting something out there that can be added to CI as a test).

**Comment by kent-mcleod** [comment]
> I assume the pc99 failure is spurious in this case.

Yea, the machine runner failed to boot and the test timed out.

**Comment by lsf37** [comment]
> I'm not sure what the current state of the CI is though after the recent rapid migration. @lsf37 would know though?

The simulation tests for that one are working, and the hardware tests are still on my list, but are definitely planned to be revived.

**Comment by lsf37** [comment]
So, @kent-mcleod, do I understand correctly that the `sylvain/gic-v3` is the updated version of #145? Tbh I'd prefer that over the current PR since it integrates a bit further. 

Do you want to make it into a PR?

**Comment by kent-mcleod** [comment]
> Do you want to make it into a PR?

I can update the old PR with it.

**Comment by kent-mcleod** [comment]
> I don't have a strong feeling one way or the other. I've got zero problem ditching this PR if someone wants to do the work to get #145 ready for merge. I've also got zero problems with getting this merged now, and then applying the refactors that are in #145.
> 
> I think getting _something_ in sooner rather than later is beneficial as it allows others to build on this (including potentially getting something out there that can be added to CI as a test).

Yea, I don't want to end up going in circles and not make forward progress either.  We'd at least need either you and/or @Indanz to confirm that the patch set in the other PR is sufficient for you to use though.

File: include/arch/arm/arch/machine/gic_v3.h
Comment (kent-mcleod) [reply include/arch/arm/arch/machine/gic_v3.h:424]: The other PR addresses this

Comment (kent-mcleod) [reply include/arch/arm/arch/machine/gic_v3.h:424]: The GIC_VCPU_MAX_NUM_LR is overridden by the global gic_vcpu_num_list_regs which is initialized during boot to the actual number that the hardware supports.  So this would still work if GIC_VCPU_MAX_NUM_LR is set to 64 as that would just impact the size of the struct in the VCPU object.

**Comment by Indanz** [comment]
Can anyone explain why  #145 is still not merged after two years while there was working CI? Just the lack of reviewing or is there something more fundamental going on that prevents PR's from getting merged, like it being unclear who's responsible for what? SeL4 needs a decision making process that guarantees forward progress for PRs. It is fine to keep them open if it's unclear if the feature is wanted or if it needs verification. But it should be clear to everyone what's blocking the merge and what the next step should be.

In this case it's extra worrisome because it's a feature that's wanted and has no verification impact. If even Anna Lyons has difficulty getting PRs merged, what hope do others have?

I don't care which PR gets merged, as long as seL4 gets virtual GICv3 support. The code in this PR has been working for me for more than half a year now. I assume Ben had a reason not to use PR #145. I can test the updated #145 too, but that is easier to do if the TQMa8XQP platform support has been merged.

**Comment by bennoleslie** [comment]
Regarding why I didn't use PR #145 in the first place. At the time I could not get that working out of the box and it wasn't clear exactly what the state of PR #145 was.

The other reason was that this approach is just a simpler diff. With one exception it only adds new code, and doesn't rework existing code. As I was maintaining this patch out of tree for sometime having a simple diff was very beneficial to me.

To be very clear: none of this is meant to be critical of #145 ! I was just working from different constraints. If #145 can be updated and merged now then I've got no problem.

**Comment by lsf37** [comment]
> Can anyone explain why #145 is still not merged after two years while there was working CI? Just the lack of reviewing or is there something more fundamental going on that prevents PR's from getting merged, like it being unclear who's responsible for what? 

For the situation from 2 years ago, I think I can explain that fairly easily: too few people with too much to do. Much of the development work at that time was not visible publicly and for contracts with deadlines which tends to crowd out other work. That situation got worse continually until it pretty much exploded earlier this year.

While the D61 group now does not exist any more and the systems group at UNSW is  smaller, the seL4 foundation now has full control of the repositories and the contribution process, and reviewing is done almost 100% by volunteers such as myself, @kent-mcleod, @yyshen, @axel-h, you and others. Together with the move of all development into the open to GitHub, this has increased responsiveness a lot AFAICS. There is a big backlog, but it is shrinking. Which is one reason we're getting back to #145

The foundation now also has some budget to spend on these issues, so maybe we will see further improvement from that direction.

> SeL4 needs a decision making process that guarantees forward progress for PRs. It is fine to keep them open if it's unclear if the feature is wanted or if it needs verification. But it should be clear to everyone what's blocking the merge and what the next step should be.

The decision making process is clear -- a change needs to be an improvement, needs to come with verification or not impact verification, needs to pass tests or if it implements new features needs to come with new tests, and it needs to fit with the design of seL4. If it changes the API it needs a larger discussion and RFC. Other considerations are added maintenance burden or not. What is delaying most merges is reviewer bandwidth. The more people help with that, the faster we can move. Currently, CI is problematic, but that will be temporary. 

GICv3 virtualisation support is clearly an improvement and feature that is wanted. We just need to make it happen.


> In this case it's extra worrisome because it's a feature that's wanted and has no verification impact. 

It turns out #145 does have verification impact, hopefully minor. I'm doing that proof update for free, taking time out from my own company, but I don't think it will take more than a few days to figure out the details. If it turns out that it is a too-large job, we should probably merge this PR instead, but I'd like to give #145 a shot first.

**Comment by kent-mcleod** [comment]
> Can anyone explain why #145 is still not merged after two years while there was working CI?

There was no working CI that could test the GICv3 virtualization code though as the camkes-vm project is missing a user level driver for GICv3 emulation. @pingerino was working on this 2 years ago and got to a point where the an SMP linux guest VM using the GICv3 would boot to login on the imx8mq-evk platform.  But before she could cleanup her implementation and get it merged, she had to abandon it as work-in-progress and left it as WIP PRs on the relevant external repos for anyone else to pick up and complete: (https://github.com/seL4/seL4/pull/145, https://github.com/seL4/seL4_projects_libs/pull/8, https://github.com/seL4/camkes-vm-images/pull/2, https://github.com/seL4/camkes-arm-vm/pull/6)

While the changes in the kernel are quite minimal, the changes to add the userlevel emulation driver where more substantial, and without anyone actually pushing to get the change merged it proceeded to bitrot faster than there was spare cycles available within Data61 to continue with the cleanup and review process.

> But it should be clear to everyone what's blocking the merge and what the next step should be

In this case the old PR has been open for so long because there was no-one working consistently enough on finishing off the camkes-vm project support and no one pushing for the kernel changes to be urgently merged.

Now that both @Indanz and @bennoleslie are advocating for this PR to get merged, it should hopefully happen pretty quickly.

**Comment by kent-mcleod** [comment]
> I can test the updated #145 too, but that is easier to do if the TQMa8XQP platform support has been merged.

It's probably not worth it because of https://github.com/seL4/seL4/pull/145#issuecomment-908191870.  After going through #145 and reviewing it for merge, it looks like a lot of the refactoring common code changes are due to the GICv3 implementation in that PR managing maintenance interrupts quite differently to the v2. Instead of sending a single IRQ in a VGICMaintenance fault like the GICv2, it sends multiple by bitwise orring them into the irq parameter.  If we don't make this change, then there's essentially no difference from this PR.

**Comment by Indanz** [comment]
Thanks for the clear explanations!

We just need virtual GICv3 support and are not using camkes-vm. Ben wrote a VMM which we are using, that could be used for CI testing instead of camkes if that simplifies things.

**Comment by yyshen** [comment]
> > Can anyone explain why #145 is still not merged after two years while there was working CI? Just the lack of reviewing or is there something more fundamental going on that prevents PR's from getting merged, like it being unclear who's responsible for what?
> 
> For the situation from 2 years ago, I think I can explain that fairly easily: too few people with too much to do. Much of the development work at that time was not visible publicly and for contracts with deadlines which tends to crowd out other work. That situation got worse continually until it pretty much exploded earlier this year.
> 
> While the D61 group now does not exist any more and the systems group at UNSW is smaller, the seL4 foundation now has full control of the repositories and the contribution process, and reviewing is done almost 100% by volunteers such as myself, @kent-mcleod, @yyshen, @axel-h, you and others. Together with the move of all development into the open to GitHub, this has increased responsiveness a lot AFAICS. There is a big backlog, but it is shrinking. Which is one reason we're getting back to #145
> 
> The foundation now also has some budget to spend on these issues, so maybe we will see further improvement from that direction.
> 
> > SeL4 needs a decision making process that guarantees forward progress for PRs. It is fine to keep them open if it's unclear if the feature is wanted or if it needs verification. But it should be clear to everyone what's blocking the merge and what the next step should be.
> 
> The decision making process is clear -- a change needs to be an improvement, needs to come with verification or not impact verification, needs to pass tests or if it implements new features needs to come with new tests, and it needs to fit with the design of seL4. If it changes the API it needs a larger discussion and RFC. Other considerations are added maintenance burden or not. What is delaying most merges is reviewer bandwidth. The more people help with that, the faster we can move. Currently, CI is problematic, but that will be temporary.
> 
> GICv3 virtualisation support is clearly an improvement and feature that is wanted. We just need to make it happen.
> 
> > In this case it's extra worrisome because it's a feature that's wanted and has no verification impact.
> 
> It turns out #145 does have verification impact, hopefully minor. I'm doing that proof update for free, taking time out from my own company, but I don't think it will take more than a few days to figure out the details. If it turns out that it is a too-large job, we should probably merge this PR instead, but I'd like to give #145 a shot first.

@lsf37 The behaviours I observed on some hardware might not exhibit on the imx8 platform.  Still, it is something we might need to deal with in the future if more GICv3-based platforms are supported.

**Comment by yyshen** [comment]
> > I can test the updated #145 too, but that is easier to do if the TQMa8XQP platform support has been merged.
> 
> It's probably not worth it because of [#145 (comment)](https://github.com/seL4/seL4/pull/145#issuecomment-908191870). After going through #145 and reviewing it for merge, it looks like a lot of the refactoring common code changes are due to the GICv3 implementation in that PR managing maintenance interrupts quite differently to the v2. Instead of sending a single IRQ in a VGICMaintenance fault like the GICv2, it sends multiple by bitwise orring them into the irq parameter. If we don't make this change, then there's essentially no difference from this PR.

@bennoleslie Have you observed that one maintenance interrupt represents two EOIs on the imx8 board? I guess no?

**Comment by kent-mcleod** [comment]
Ok, I made https://github.com/BreakawayConsulting/seL4/pull/1 which is my attempt of taking the refactoring that we want from https://github.com/seL4/seL4/pull/145 and merging it with this PR

File: include/arch/arm/arch/machine/gic_v3.h
Comment (kent-mcleod) [inline include/arch/arm/arch/machine/gic_v3.h:381]: I think this can work because we leave the max priority for decodeVCPUInjectIRQ as 31.

**Comment by kent-mcleod** [comment]
> Ok, I made [BreakawayConsulting#1](https://github.com/BreakawayConsulting/seL4/pull/1) which is my attempt of taking the refactoring that we want from #145 and merging it with this PR

@bennoleslie are you happy to add these changes into this PR branch and test it with your VMM before merging.  There shouldn't be any behavior changes, just doing the bitfield extraction from the older PR and updating the GIC_VCPU_MAX_NUM_LR definition

**Comment by bennoleslie** [comment]
With the tqma8xqp1gb I can at least run tests with `-DARM_HYP=ON` for a supported platform. I don't think it really exercises the functionality as such, but it ensures it doesn't break anything. At a minimum it gets us a buildable/runnable kernel with hyp enabled.

Results are:

* `-DPLATFORM=tqma8xqp1gb -DRELEASE=OFF -DMCS=OFF -DSMP=OFF -DARM_HYP=ON` (129 passed, 41 disabled)
* `-DPLATFORM=tqma8xqp1gb -DRELEASE=ON -DMCS=OFF -DSMP=OFF -DARM_HYP=ON` (129 passed, 41 disabled)
* `-DPLATFORM=tqma8xqp1gb -DRELEASE=OFF -DMCS=ON -DSMP=OFF -DARM_HYP=ON` (175 passed, 14 disabled)
* `-DPLATFORM=tqma8xqp1gb -DRELEASE=ON -DMCS=ON -DSMP=OFF -DARM_HYP=ON` (175 passed, 14 disabled)
* `-DPLATFORM=tqma8xqp1gb -DRELEASE=OFF -DMCS=OFF -DSMP=ON -DARM_HYP=ON` (134 passed,  36 disabled)
* `-DPLATFORM=tqma8xqp1gb -DRELEASE=ON -DMCS=OFF -DSMP=ON -DARM_HYP=ON` (134 passed,  36 disabled)
* `-DPLATFORM=tqma8xqp1gb -DRELEASE=OFF -DMCS=ON -DSMP=ON -DARM_HYP=ON` (181 passed, 8 disabled)
* `-DPLATFORM=tqma8xqp1gb -DRELEASE=ON -DMCS=ON -DSMP=ON -DARM_HYP=ON` (181 passed,  8 disabled)

Note: I'm still happy with either #145 or this one. I'm just putting this here as an option.

**Comment by bennoleslie** [comment]
> > Ok, I made [BreakawayConsulting#1](https://github.com/BreakawayConsulting/seL4/pull/1) which is my attempt of taking the refactoring that we want from #145 and merging it with this PR
> 
> @bennoleslie are you happy to add these changes into this PR branch and test it with your VMM before merging. There shouldn't be any behavior changes, just doing the bitfield extraction from the older PR and updating the GIC_VCPU_MAX_NUM_LR definition

Sorry, missed this I was was running tests. I'm looking at your PR now, and will work out how best to test it.

**Comment by kent-mcleod** [comment]
I meant more test that it works in your hypervisor implementation if you are happy to. As we don't have any tests we can run ourselves.

**Comment by bennoleslie** [comment]
> I meant more test that it works in your hypervisor implementation if you are happy to. As we don't have any tests we can run ourselves.

Yes, sorry, for the confusion. Overlapping messages at slightly different times :).

I'd started that testing above *before* I saw your messages about the updates you an @yyshen had made. 

I will test the branch with my  the current VMM project. I need to do some boring admin stuff there first (bring all the repos up-to-date and so forth). Unfortunately I probably won't get that done until first this in the morning (Sydney time).

**Comment by bennoleslie** [comment]
> > I meant more test that it works in your hypervisor implementation if you are happy to. As we don't have any tests we can run ourselves.
> 
> Yes, sorry, for the confusion. Overlapping messages at slightly different times :).
> 
> I'd started that testing above _before_ I saw your messages about the updates you an @yyshen had made.
> 
> I will test the branch with my the current VMM project. I need to do some boring admin stuff there first (bring all the repos up-to-date and so forth). Unfortunately I probably won't get that done until first this in the morning (Sydney time).

FYI: I'm testing now. Unfortunately it appears that other seL4_tools or util_libs changes have broken what was previously working for me, so I need to sort that out.

**Comment by kent-mcleod** [comment]
@lsf37 Are you happy that those preprocess errors are insignificant?

**Comment by lsf37** [comment]
> @lsf37 Are you happy that those preprocess errors are insignificant?

Let me run the proofs, it's possible that they are affected.

**Comment by bennoleslie** [comment]
I've cherry picked @yyshen and @kent-mcleod suggested patches into the branch, and rebased onto latest master.

I've tested this in my vmm project and can successfully boot Linux on all ON/OFF combinations of RELEASE/MCS/SMP.

If others are happy I think we could merge.

**Comment by bennoleslie** [comment]
@kent-mcleod Can you please have a look at the DCO failure. It seems upset that you signed off, but the commit author was @yyshen  I don't understand how best to deal with that.

**Comment by kent-mcleod** [comment]
> @kent-mcleod Can you please have a look at the DCO failure. It seems upset that you signed off, but the commit author was @yyshen I don't understand how best to deal with that.

@lsf37 is there a way that we can turn off that part of the check?  It just needs to enforce that the commit is signed off right?

@bennoleslie it's not an issue, we just manually check the test to passing in this case.

**Comment by bennoleslie** [comment]
Well, the proof failure is disappointing.

Is there anything I can do on the code side to try and solve that? I'm afraid I couldn't really understand what was going on with the errors.

File: src/arch/arm/config.cmake
Comment (Indanz) [inline src/arch/arm/config.cmake:264]: Shouldn't this be "KernelArmHypervisorSupport;KernelArchARM"?

For GICv3 it perhaps makes more sense to test for KernelArmGicV3 directly? And if so, perhaps test for KernelArmGicV2 here as well then?

Comment (bennoleslie) [reply src/arch/arm/config.cmake:264]: I think the initial patch was more explicit around using KernelArmGicV3, but the refactor patches from @yyshen and @kent-mcleod preferred it this way. I think there is probably another critical look at what is 'GIC' and what is 'GIC extension for virtualization' and where that code belongs. But I really wanted to get the simplest thing that people would agree on in the first instance.

**Comment by lsf37** [comment]
> Well, the proof failure is disappointing.
> 
> Is there anything I can do on the code side to try and solve that? I'm afraid I couldn't really understand what was going on with the errors.

Having a look now. It's something with the re-organised bitfield structures. Might be trivial, but I can't really tell from the log either, need to stare at it in the prover.

> @lsf37 is there a way that we can turn off that part of the check? It just needs to enforce that the commit is signed off right?

Unfortunately it looks like there is not much we can do about it. The app source is [here](https://github.com/probot/dco) and the only config option allows you to waiver the sign-off from org members who sign with gpg instead. We could fork the app, add something and host it ourselves, but it's probably easier to just set it to "pass" every once in a while.

File: src/arch/arm/config.cmake
Comment (kent-mcleod) [reply src/arch/arm/config.cmake:264]: This guard is whether a header gets generated for the file.  By not guarding for KernelArmHypervisorSupport the header will always get generated, but it only ever gets included during hypervisor support.  We can make them consistent, but it didn't really reach my threshold for importance when merging patches together.

File: src/plat/tqma8xqp1gb/config.cmake
Comment (kent-mcleod) [inline src/plat/tqma8xqp1gb/config.cmake:15]: We should remove this as it's from before the config was removed.

**Comment by kent-mcleod** [comment]
> Having a look now. It's something with the re-organised bitfield structures. Might be trivial, but I can't really tell from the log either, need to stare at it in the prover.

The only thing that is supposed to have changed is the definition's position in the final AST...

File: src/plat/tqma8xqp1gb/config.cmake
Comment (bennoleslie) [reply src/plat/tqma8xqp1gb/config.cmake:15]: Agreed. That is a left over. I'll wait until @lsf37 is able to say something about the proof impacts before I make any additional changes.

**Comment by lsf37** [comment]
> > Having a look now. It's something with the re-organised bitfield structures. Might be trivial, but I can't really tell from the log either, need to stare at it in the prover.
> 
> The only thing that is supposed to have changed is the definition's position in the final AST...

Order might change some internal naming (not supposed to do that any more, but I doubt we are catching all cases). If that it the case it should be very easy to fix. It's building now, but will take 2h or so until it's there and I can say more.

**Comment by lsf37** [comment]
> Order might change some internal naming (not supposed to do that any more, but I doubt we are catching all cases). If that it the case it should be very easy to fix. It's building now, but will take 2h or so until it's there and I can say more.

I found the proof breakage: it's just that the auto-generated bitfield proofs for `gic_v2.bf` end up in a new file that needs to be included, which is a trivial two-line change. Currently building, but should be all good.

**Comment by bennoleslie** [comment]
> > Order might change some internal naming (not supposed to do that any more, but I doubt we are catching all cases). If that it the case it should be very easy to fix. It's building now, but will take 2h or so until it's there and I can say more.
> 
> I found the proof breakage: it's just that the auto-generated bitfield proofs for `gic_v2.bf` end up in a new file that needs to be included, which is a trivial two-line change. Currently building, but should be all good.

I'll make those other suggested changes and squash/push.

I'm not quite sure how the things work from here. I guess the proof changes are on a different branch?

**Comment by lsf37** [comment]
> I'm not quite sure how the things work from here. I guess the proof changes are on a different branch?

I'll make a separate PR on l4v for the proof changes. When that is up, we can merge both.

**Comment by kent-mcleod** [comment]
Then shouldn't we just set up the includes on the .bf side?

**Comment by bennoleslie** [comment]
> Then shouldn't we just set up the includes on the .bf side?

Sorry, can you explain in more detail?

**Comment by kent-mcleod** [comment]
Basically drop the commit that undoes your build system changes and creates the secondary .bf files .....

**Comment by bennoleslie** [comment]
> Basically drop the commit that undoes your build system changes and creates the secondary .bf files .....

OK, easy. I'll do that now.

**Comment by lsf37** [comment]
Corresponding PR seL4/l4v#349 is now up

**Comment by bennoleslie** [comment]
Per @kent-mcleod suggestion I dropped the commit that split thing into multiple .bf files.

But I'm now confused if the verification change is required. Sorry, maybe we need to coordinate in a more real-time manner.

**Comment by lsf37** [comment]
> Per @kent-mcleod suggestion I dropped the commit that split thing into multiple .bf files.
> 
> But I'm now confused if the verification change is required. Sorry, maybe we need to coordinate in a more real-time manner.

With that we don't need a verification change, and I can drop the l4v PR. That's also fine, no worries. In fact, that's slightly nicer, so I'd prefer it.

**Comment by bennoleslie** [comment]
> > Per @kent-mcleod suggestion I dropped the commit that split thing into multiple .bf files.
> > But I'm now confused if the verification change is required. Sorry, maybe we need to coordinate in a more real-time manner.
> 
> With that we don't need a verification change, and I can drop the l4v PR. That's also fine, no worries. In fact, that's slightly nicer, so I'd prefer it.

OK, cool. Sorry for the busy work.

**Comment by bennoleslie** [comment]
I'm pretty sure that the pc99 failure is spurious. I think this means things are good to merge?

**Review by kent-mcleod — APPROVED** [review]

**Comment by kent-mcleod** [comment]
Thanks for your patience :)

**Comment by bennoleslie** [comment]
Thanks @Indanz @kent-mcleod @lsf37 @pingerino @yyshen @sylGauthier for directly or indirectly helping get this functionality into the kernel! (Sorry if I missed anyone!)

**Comment by yyshen** [comment]
> Thanks @Indanz @kent-mcleod @lsf37 @pingerino @yyshen @sylGauthier for directly or indirectly helping get this functionality into the kernel! (Sorry if I missed anyone!)

Very glad to see the function is finally merged. Thanks everyone!

**Comment by Indanz** [comment]
Hurray! Thanks for your time everyone!

**Pull request merged**
